File: BOJ/BOJ15683.java
package com.ssafy.algo;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class BOJ15683 {
	static int N, M, cnt, cctv, minn = 64;
	static int[][] office;
	static int[][] temp;
	static ArrayList<info> arr = new ArrayList<info>();
	
	static class info{
		int c;
		int r;
		int num;
		int dir;
		
		public info() {
			
		}

		public info(int c, int r, int num, int dir) {
			super();
			this.c = c;
			this.r = r;
			this.num = num;
			this.dir = dir;
		}
	}

	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        office = new int[N][M];
        temp = new int[N][M];
        
    	for (int i = 0; i < N; i++) {
    		st = new StringTokenizer(br.readLine());
    		for (int j = 0; j < M; j++) {
    			office[i][j] = Integer.parseInt(st.nextToken());
    			if (office[i][j] != 0 && office[i][j] != 6) {
    				arr.add(new info(i,j,office[i][j], -1));
    				cctv++;
    			}
    		}
    	}
    	
    	solve(0);
    	bw.write(minn + "\n");
    	br.close();
    	bw.flush();
    	bw.close();
	}
	public static void solve(int count) {
		if (count == cctv) {
			check();
			minn = Math.min(minn, cal());
			return;
		}

		arr.get(count).dir= 0;
		solve(count + 1);

		arr.get(count).dir = 1;
		solve(count + 1);

		arr.get(count).dir = 2;
		solve(count + 1);

		arr.get(count).dir = 3;
		solve(count + 1);

	}
	
	public static void check(){
		reset();
		for (int i = 0; i < cctv; i++) {
			if (arr.get(i).num == 1) {
				if (arr.get(i).dir == 0) {
					east(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 1) {
					west(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 2) {
					south(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 3) {
					north(arr.get(i).c, arr.get(i).r);
				}

			}
			else if (arr.get(i).num == 2) {
				if (arr.get(i).dir == 0 || arr.get(i).dir == 1) {
					east(arr.get(i).c, arr.get(i).r);
					west(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 2 || arr.get(i).dir == 3) {
					south(arr.get(i).c, arr.get(i).r);
					north(arr.get(i).c, arr.get(i).r);
				}

			}
			else if (arr.get(i).num == 3) {
				if (arr.get(i).dir == 0) {
					east(arr.get(i).c, arr.get(i).r);
					north(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 1) {
					west(arr.get(i).c, arr.get(i).r);
					south(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 2) {
					south(arr.get(i).c, arr.get(i).r);
					east(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 3) {
					north(arr.get(i).c, arr.get(i).r);
					west(arr.get(i).c, arr.get(i).r);
				}

			}
			else if (arr.get(i).num == 4) {
				if (arr.get(i).dir == 0) {
					east(arr.get(i).c, arr.get(i).r);
					west(arr.get(i).c, arr.get(i).r);
					north(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 1) {
					west(arr.get(i).c, arr.get(i).r);
					east(arr.get(i).c, arr.get(i).r);
					south(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 2) {
					south(arr.get(i).c, arr.get(i).r);
					north(arr.get(i).c, arr.get(i).r);
					west(arr.get(i).c, arr.get(i).r);
				}
				else if (arr.get(i).dir == 3) {
					south(arr.get(i).c, arr.get(i).r);
					north(arr.get(i).c, arr.get(i).r);
					east(arr.get(i).c, arr.get(i).r);
				}

			}
			else if (arr.get(i).num == 5) {
				east(arr.get(i).c, arr.get(i).r);
				west(arr.get(i).c, arr.get(i).r);
				south(arr.get(i).c, arr.get(i).r);
				north(arr.get(i).c, arr.get(i).r);
			}
		}
	}

	public static int cal() {
		int result = 0;

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (temp[i][j] == 0) {
					result++;
				}
			}
		}

		return result;
	}
	
	public static void reset() {
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				temp[i][j] = office[i][j];
			}
		}
	}
	
	public static void east(int x, int y) {
		int b = y;
		while (y < M - 1) {
			y++;
			if (temp[x][y] == 6) {
				break;
			}
			else if (temp[x][y] == 0) {
				temp[x][y] = -1;
			}
		}
		y = b;
	}
	
	public static void west(int x, int y) {
		int b = y;
		while (y > 0) {
			y--;
			if (temp[x][y] == 6) {
				break;
			}
			else if (temp[x][y] == 0) {
				temp[x][y] = -1;
			}
		}
		y = b;
	}
	
	public static void south(int x, int y) {
		int a = x;
		while (x < N - 1) {
			x++;
			if (temp[x][y] == 6) {
				break;
			}
			else if (temp[x][y] == 0) {
				temp[x][y] = -1;
			}
		}
		x = a;
	}
	
	public static void north(int x, int y) {
		int a = x;
		while (x > 0) {
			x--;
			if (temp[x][y] == 6) {
				break;
			}
			else if (temp[x][y] == 0) {
				temp[x][y] = -1;
			}
		}
		x = a;
	}
}

File: BOJ/BOJ15683.cpp
//#include<iostream>
//#include<algorithm>
//#include<vector>
//#include<utility>
//
//using namespace std;
//
//int N, M;
//int cnt = 0;
//int minn = 64;
//int cctv = 0;
//int office[8][8], temp[8][8];
//vector <pair <pair <int, int>, pair<int, int> > > info;
//
//int reset() {
//	for (int i = 0; i < N; i++) {
//		for (int j = 0; j < M; j++) {
//			temp[i][j] = office[i][j];
//		}
//	}
//
//	return 0;
//}
//
//int east(int x, int y) {
//	int b = y;
//	while (y < M - 1) {
//		y++;
//		if (temp[x][y] == 6) {
//			break;
//		}
//		else if (temp[x][y] == 0) {
//			temp[x][y] = -1;
//		}
//	}
//	y = b;
//
//	return 0;
//}
//
//int west(int x, int y) {
//	int b = y;
//	while (y > 0) {
//		y--;
//		if (temp[x][y] == 6) {
//			break;
//		}
//		else if (temp[x][y] == 0) {
//			temp[x][y] = -1;
//		}
//	}
//	y = b;
//
//	return 0;
//}
//
//int south(int x, int y) {
//	int a = x;
//	while (x < N - 1) {
//		x++;
//		if (temp[x][y] == 6) {
//			break;
//		}
//		else if (temp[x][y] == 0) {
//			temp[x][y] = -1;
//		}
//	}
//	x = a;
//
//	return 0;
//}
//
//int north(int x, int y) {
//	int a = x;
//	while (x > 0) {
//		x--;
//		if (temp[x][y] == 6) {
//			break;
//		}
//		else if (temp[x][y] == 0) {
//			temp[x][y] = -1;
//		}
//	}
//	x = a;
//
//	return 0;
//}
//
//int check() {
//	reset();
//	for (int i = 0; i < cctv; i++) {
//		if (info[i].second.first == 1) {
//			if (info[i].second.second == 0) {
//				east(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 1) {
//				west(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 2) {
//				south(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 3) {
//				north(info[i].first.first, info[i].first.second);
//			}
//
//		}
//		else if (info[i].second.first == 2) {
//			if (info[i].second.second == 0 || info[i].second.second == 1) {
//				east(info[i].first.first, info[i].first.second);
//				west(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 2 || info[i].second.second == 3) {
//				south(info[i].first.first, info[i].first.second);
//				north(info[i].first.first, info[i].first.second);
//			}
//
//		}
//		else if (info[i].second.first == 3) {
//			if (info[i].second.second == 0) {
//				east(info[i].first.first, info[i].first.second);
//				north(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 1) {
//				west(info[i].first.first, info[i].first.second);
//				south(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 2) {
//				south(info[i].first.first, info[i].first.second);
//				east(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 3) {
//				north(info[i].first.first, info[i].first.second);
//				west(info[i].first.first, info[i].first.second);
//			}
//
//		}
//		else if (info[i].second.first == 4) {
//			if (info[i].second.second == 0) {
//				east(info[i].first.first, info[i].first.second);
//				west(info[i].first.first, info[i].first.second);
//				north(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 1) {
//				west(info[i].first.first, info[i].first.second);
//				east(info[i].first.first, info[i].first.second);
//				south(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 2) {
//				south(info[i].first.first, info[i].first.second);
//				north(info[i].first.first, info[i].first.second);
//				west(info[i].first.first, info[i].first.second);
//			}
//			else if (info[i].second.second == 3) {
//				south(info[i].first.first, info[i].first.second);
//				north(info[i].first.first, info[i].first.second);
//				east(info[i].first.first, info[i].first.second);
//			}
//
//		}
//		else if (info[i].second.first == 5) {
//			east(info[i].first.first, info[i].first.second);
//			west(info[i].first.first, info[i].first.second);
//			south(info[i].first.first, info[i].first.second);
//			north(info[i].first.first, info[i].first.second);
//		}
//	}
//
//	return 0;
//}
//
//int cal() {
//	int result = 0;
//
//	for (int i = 0; i < N; i++) {
//		for (int j = 0; j < M; j++) {
//			if (temp[i][j] == 0) {
//				result++;
//			}
//		}
//	}
//
//	return result;
//}
//
//int solve(int count) {
//	if (count == cctv) {
//		check();
//		minn = min(minn, cal());
//		return 0;
//	}
//
//	info[count].second.second = 0;
//	solve(count + 1);
//
//	info[count].second.second = 1;
//	solve(count + 1);
//
//	info[count].second.second = 2;
//	solve(count + 1);
//
//	info[count].second.second = 3;
//	solve(count + 1);
//
//
//	return 0;
//}
//
//int main() {
//	ios_base::sync_with_stdio(false);
//	cin.tie(0);
//	cout.tie(0);
//
//	cin >> N >> M;
//
//	for (int i = 0; i < N; i++) {
//		for (int j = 0; j < M; j++) {
//			cin >> office[i][j];
//			if (office[i][j] != 0 && office[i][j] != 6) {
//				info.push_back(make_pair(make_pair(i, j), make_pair(office[i][j], -1)));
//				cctv++;
//			}
//		}
//	}
//
//	solve(0);
//
//	cout << minn << "\n";
//
//	return 0;
//}